src/simulation/dynamics/_GeneralModuleFiles/svIntegratorAdaptiveRungeKutta.h

```diff
@@ -0,0 +1,371 @@
+ */
+template <size_t numberStages> struct RKAdaptiveCoefficients : public RKCoefficients<numberStages> {
+
+    /** "b*" Coefficients of the Adaptive RK method ("b" coefficients of the higher order method) */
+    typename RKCoefficients<numberStages>::StageSizedArray bStarArray = {};
+};
+
+/**
+ * The svIntegratorRungeKutta class implements a state integrator based on the
+ * family of explicit Runge-Kutta numerical integrators with variable time step.
+ *
+ * When 'svIntegratorAdaptiveRungeKutta::integrate' is called, this integrator will
+ * try to integrate with the given time step. It will then evaluate the error commited
+ * and, if it's too large, internally use smaller time steps until the error tolerances
+ * are met.
+ */
+template <size_t numberStages>
+class svIntegratorAdaptiveRungeKutta : public svIntegratorRungeKutta<numberStages> {
+  public:
+    /**
+     * Constructs the integrator for the given dynamic object and specified Adaptive RK
+     * coefficients.
+     *
+     * methodLargestOrder is the higher order of the two
+     * orders used in adaptive RK methods. For the RKF45 method,
+     * for example, methodLargestOrder should be 5.
+     */
+    svIntegratorAdaptiveRungeKutta(DynamicObject* dynIn,
+                                   const RKAdaptiveCoefficients<numberStages>& coefficients,
+                                   const double methodLargestOrder);
+
+    /** Performs the integration of the associated dynamic objects up to time currentTime+timeStep
+     */
+    virtual void integrate(double currentTime, double timeStep) override;
+
+    /**
+     * Sets the relative tolerance for every DynamicObject
+     * and every state.
+     *
+     * This is akin to changing svIntegratorAdaptiveRungeKutta::relTol
+     * directly
+     */
+    void setRelativeTolerance(double relTol);
+
+    /**
+     * Sets the absolute tolerance for every DynamicObject
+     * and every state.
+     *
+     * This is akin to changing svIntegratorAdaptiveRungeKutta::absTol
+     * directly
+     */
+    void setAbsoluteTolerance(double absTol);
+
+    /**
+     * Sets the relative tolerance for every DynamicObject
+     * and the state identified by the given name.
+     */
+    void setRelativeTolerance(std::string stateName, double relTol);
+
+    /**
+     * Sets the absolute tolerance for every DynamicObject
+     * and the state identified by the given name.
+     */
+    void setAbsoluteTolerance(std::string stateName, double absTol);
+
+    /**
+     * Sets the relative tolerance for the given DynamicObject
+     * and the state identified by the given name.
+     */
+    void setRelativeTolerance(const DynamicObject& dynamicObject, std::string stateName,
+                              double relTol);
+
+    /**
+     * Sets the absolute tolerance for the given DynamicObject
+     * and the state identified by the given name.
+     */
+    void setAbsoluteTolerance(const DynamicObject& dynamicObject, std::string stateName,
```

Comment:
Yes, I would say that that is useful. However, how do we handle multiple conflicting tolerances? Right now, you can define tolerances from less specific to more specific: general -> state -> dyn obj + state. Then, the most specific tolerance is used. If we were to add DynamicObject tolerances, they would live in the same specificity level as "state". In that case, which one to choose? The smallest? It may be confusing. Thoughts?

Reply:
Wouldn't this tolerance live between general (integrator) and state (specific quantity to be integrated)?

Reply:
Well, if you set the tolerance for the state, then it applies to every DynamicObject for that particular state. If you set it for a DynamicObject, it applies to every state for that particular DynamicObject. To me, it is not clear which one of these statements is more specific and thus should take priority.

If you set the tolerance for state "hubPosition" to be 1e-4 and for the DynamicObject "debris" to be 1e-5, what is the tolerance that should be used for "hubPosition" for "debris"?

Reply:
I see what you're saying. The state tolerance takes priority over the DynamicObject tolerance, as it is more specific. I think this would fit what I said about it living between general and state. Thoughts?